### PR TITLE
feat(ABCI): New Proposal Struct with Associated Metadata

### PR DESCRIPTION
**Pull request by davidterpay**
## Overview
This PR creates a new proposal type that replace the current proposal building object. The premise here is the following:

By attaching metadata to the proposal about how many transactions were included by each lane, we can deprecate the `CheckOrderHandler` and solely have a single `ProcessLane` handler. Given a constructed proposal with meta data attached to it - lane x has 1 txs, lane y has 2 txs, etc. - we no longer have to do any introspection into what transactions belong to what lane. Definitively, we can take the proposal and break it up into its corresponding partial proposals - the proposals each lane built - and have stronger guarantees about the properties of the proposed block.

Additionally, this will both decrease the complexity of the code base and will allow us to subsequently do more nuanced things when building/verifying proposals. For example, this can include 
- a app hash check point between each lane to ensure that there is determinism amongst all of the lanes
- more flexibility for extending proposals to include vote extensions, and more

### CODE REVIEW

File: block/proposals/update.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	partialProposalSize := int64(0)
+	partialProposalGasLimit := uint64(0)
+
+	for index, tx := range partialProposal {
+		txInfo, err := utils.GetTxInfo(p.TxEncoder, tx)
+		if err != nil {
+			return fmt.Errorf("err retriveing transaction info: %s", err)
+		}
+
+		// Invarient check: Ensure that the transaction is not already in the proposal.
```
Comment (nivasan1) [inline block/proposals/update.go:37]: Invariant

File: block/proposals/update.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	// Aggregate info from the transactions.
+	hashes := make(map[string]struct{})
+	txs := make([][]byte, len(partialProposal))
+	partialProposalSize := int64(0)
+	partialProposalGasLimit := uint64(0)
+
+	for index, tx := range partialProposal {
+		txInfo, err := utils.GetTxInfo(p.TxEncoder, tx)
+		if err != nil {
+			return fmt.Errorf("err retriveing transaction info: %s", err)
```
Comment (nivasan1) [inline block/proposals/update.go:34]: retrieve

File: block/proposals/update.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if updatedGasLimit > p.Info.MaxGasLimit {
+		return fmt.Errorf(
+			"block proposal consumes too much gas: %d > %d",
+			updatedGasLimit,
+			p.Info.MaxGasLimit,
+		)
+	}
+
+	// Invarient check: Ensure we have not already prepared a partial proposal for this lane.
+	if _, ok := p.Info.TxsByLane[lane]; ok {
```
Comment (nivasan1) [inline block/proposals/update.go:87]: nit: would make this check the first check (avoids redundant work)

File: block/proposals/update.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+// few invarients that are checked:
+//  1. The total size of the proposal must be less than the maximum number of bytes allowed.
+//  2. The total size of the partial proposal must be less than the maximum number of bytes allowed for
+//     the lane.
+//  3. The total gas limit of the proposal must be less than the maximum gas limit allowed.
+//  4. The total gas limit of the partial proposal must be less than the maximum gas limit allowed for
+//     the lane.
+//  5. The lane must not have already prepared a partial proposal.
+//  6. The transaction must not already be in the proposal.
+func (p *Proposal) UpdateProposal(lane string, partialProposal []sdk.Tx, limit LaneLimits) error {
```
Comment (nivasan1) [inline block/proposals/update.go:20]: Why would the caller pass in their LaneLimits? Why wouldn't the proposal j get them here from the lane-string? I.e if the purpose of this abstraction is to have devs only aware of adding txs to proposals, does it not make more sense to j hide the notion of a `LaneLimit` entirely?

File: block/proposals/update.go (hunk 1 above)
Comment (davidterpay) [reply block/proposals/update.go:37]: shii, i cant spell

File: block/proposals/update.go (hunk 2 above)
Comment (davidterpay) [reply block/proposals/update.go:34]: ong i cant spell

File: block/proposals/update.go (hunk 4 above)
Comment (davidterpay) [reply block/proposals/update.go:20]: yea i agree, will update

**Review by nivasan1 — APPROVED** [review]
lgtm

**Pull request merged**